### DISCUSSION TITLE
Help Needed: Accessing Eclipse Ditto via Localhost

### DISCUSSION THREAD

**Discussion by patrycjabubiak** [eclipse-ditto/ditto#2028] [Q&A]
Hello Mr. Jaeckle, hello everyone,

I am a master's student currently working on my thesis on the topic of "Conceptualizing a Digital Twin in the Context of the Circular Economy." As part of this, I want to create a Digital Twin and use Eclipse Ditto as an open-source framework.

I followed the steps outlined on page (https://eclipse.dev/ditto/2020-04-16-connecting-to-ttn-via-mqtt.html). I installed Docker and Docker Compose and successfully cloned the Git repository. I also checked if Ditto is running and if the Ditto clusters are healthy. As far as I can tell, without much experience, everything seems normal.

However, when I try to access the URL http://localhost:8080, I get the following message in the Safari browser: "Safari can't connect to the server. Safari can't open the page 'localhost:8080' because Safari can't connect to the server 'localhost.'"

I also wondered if I need to check the Ditto API somehow and expect to receive a JSON response confirming that the API is running.

Do you have any tips on how I should proceed? I would greatly appreciate a prompt response. Thank you!

Best regards from Germany,
Patrycja
------------------------------------------------
Hallo Herr Jaeckle, hallo zusammen,

ich bin Masterstudentin und arbeite gerade an meiner Masterarbeit zum Thema "Konzeption eines digitalen Zwillings im Kontext der Kreislaufwirtschaft". Dabei möchte ich einen Digitalen Zwilling erstellen und Eclipse Ditto als Open-Source-Framework nutzen.

Ich bin den Schritten auf Seite (https://eclipse.dev/ditto/2020-04-16-connecting-to-ttn-via-mqtt.html) gefolgt: Ich habe Docker und Docker Compose installiert und das Git-Repository erfolgreich geklont. Ebenfalls habe ich überprüft, ob Ditto ausgeführt wird und ob die Ditto-Cluster gesund sind. Soweit ich das ohne jegliche Erfahrung beurteilen kann, scheint alles normal zu sein.

Wenn ich jedoch versuche, auf die URL http://localhost:8080/ zuzugreifen, erhalte ich im Safari-Browser die folgende Meldung: „Safari kann keine Verbindung zum Server aufbauen. Safari kann die Seite 'localhost:8080' nicht öffnen, da Safari keine Verbindung zum Server 'localhost' aufbauen kann.“

Ich habe mich ebenfalls gefragt, ob ich die Ditto-API irgendwie überprüfen muss und eine JSON-Antwort erhalten sollte, die bestätigt, dass die API läuft.

Hätten Sie vielleicht einen Tipp, wie ich weiter vorgehen sollte? Ich würde mich über eine baldige Antwort sehr freuen. Vielen Dank im Voraus!

Beste Grüße aus Deutschland,
Patrycja

**Comment by Reebs637** [comment]
Hello if you need help with your thesis you can contact on this number +44 7460 560474.

**Comment by kalinkostashki** [comment]
I didn't know there was an official ditto support. 
@thjaeckle is this your number? :smile_cat: 

@patrycjabubiak I will try and answer your question:
If you've installed ditto without any argments you may need to expose the Gateway service in kubernetes with a Nodeport or LoadBalancer.
For example after I've installed ditto I get the following services:

```k get svc
NAME                 TYPE           CLUSTER-IP      EXTERNAL-IP   PORT(S)          AGE
my-ditto-dittoui     ClusterIP      10.43.224.254   <none>        8080/TCP         123m
my-ditto-gateway     ClusterIP      10.43.254.193   <none>        8080/TCP         123m
my-ditto-mongodb     ClusterIP      10.43.226.61    <none>        27017/TCP        123m
my-ditto-nginx       ClusterIP      10.43.228.174   <none>        8080/TCP         123m
my-ditto-swaggerui   ClusterIP      10.43.53.51     <none>        8080/TCP         123m
```

Using the expose command I can **expose** the my-ditto-nginx service to have a local ip address via a load balancer or Nodeport which will allocate a port on localhost that will lead to the nginx service:

```
kubectl expose svc my-ditto-nginx --type=LoadBalancer --name=cool-name
```

Which will lead to the creation of a Service type resource with a local ip address:

```
cool-name            LoadBalancer   10.43.252.32    172.18.0.2    8080:32493/TCP   2m3s
```

You can go to that IP address and you have access now :)

Another approach would be to use helm chart configs directly and create a NodePort from that:

https://github.com/eclipse-ditto/ditto/blob/405c3e1249926759755b3424415f775c61e0145e/deployment/helm/ditto/values.yaml#L1989

Hope this helped :smiley:
